### DISCUSSION TITLE
How to derive the formula in class FGInitialCondition

### DISCUSSION THREAD

**Discussion by emptyxiao** [JSBSim-Team/jsbsim#723] [Support requests]
How to derive the formula in function "calcThetaBeta"

https://github.com/JSBSim-Team/jsbsim/blob/59c0bd909dbe7232e77ac9dc03eb65f959f16cf9/src/initialization/FGInitialCondition.cpp#L321-L390

**Comment by seanmcleod** [comment]
So in the simple case assuming 0 bank angle ( $\phi$ ) and the supplied velocity vector $V$ and a requested AoA ( $\alpha$ ) the calculation is simple:  

$$ \theta = \gamma + \alpha $$ 

![image](https://user-images.githubusercontent.com/824856/190572514-02e3ec0f-a5ce-4806-a413-ac8b1186e5b2.png)

However the `calcThetaBeta(alpha, vt_NED)` function needs to handle the more general case in terms a velocity vector ( $V$ ) supplied in the NED coordinate frame, a requested $\alpha$ and orientation/attitude constraints that have been set separately, i.e. the bank angle ( $\phi$ ) and heading angle ( $\psi$ ). 

Given the constraints, in particular the orientation constraints this means that the function isn't always able to calculate a solution. For example say the bank angle is 90 degrees and the heading angle is aligned with the velocity vector's horizontal angle and a non-zero alpha is requested then it's impossible to achieve the requested alpha angle. In this case the only way to achieve the requested alpha angle would be to relax the heading angle constraint.

Here are some images drawn by @agodemar which shows all the relevant angles for the general case. Maybe they'll make it a bit clearer in terms of how the function goes about it's calculation.

https://jsbsim-team.github.io/jsbsim-reference-manual/mypages/user-manual-frames-of-reference/

![image](https://user-images.githubusercontent.com/824856/190634771-b4cd1c63-c180-478d-ae81-d1b01c316f5b.png)

![image](https://user-images.githubusercontent.com/824856/190634864-34f8137c-7d16-4b11-8657-78ac40afa8a3.png)

![image](https://user-images.githubusercontent.com/824856/190634938-be1b8404-ba05-4f26-a0c7-da9587dffe0b.png)

**Reply by emptyxiao** [reply]
Thank you for your reply! But I also want to know the mathematical derivation process. It really bothers me.
Can you demonstrate the mathematical derivation process, or provide relevant mathematical derivation process papers?
Thank you very much!

**Comment by seanmcleod** [comment]
@emptyxiao are you looking for a detailed explanation, i.e. virtually line by line explaining the maths/geometry behind every line in the function, or do you follow say 80% of the code and you just need a detailed explanation for some 20% of the code, i.e. particular steps you don't follow?

I don't have a detailed line by line explanation handy, I would need to spend a fair bit of time going through the code. 

Going through the commit history I've figured out when this particular function was added and by whom.

"Complete rewrite of the FGInitialCondition class" - https://github.com/JSBSim-Team/jsbsim/commit/da0b947502b4a74788fed1cdeac3a60a95032c3e, in 2010, method originally called `SetAlphaRadIC()`.

The good news is the code author, @bcoconni is still around and very active in terms of JSBSim.

**Reply by agodemar** [reply]
@emptyxiao 

> But I also want to know the mathematical derivation process. It really bothers me.
Can you demonstrate the mathematical derivation process, or provide relevant mathematical derivation process papers?
Thank you very much!

I can feel your thirst for learning the source of these formulas.

I think that this paper [De Marco A., Duke E. L., Berndt, J. S., "A General Solution to the Aircraft Trim Problem." AIAA Paper 2007-6703, AIAA Modeling and Simulation Technologies Conference and Exhibit 20-23 August 2007, Hilton Head, South Carolina, USA](http://wpage.unina.it/agodemar/DSV-DQV/AIAA-2007-6703-905_DeMarco_Duke_Berndt.pdf) might be of some help to you, pages from 15 to 24.

You coud help us to double check if the formulas we use in code are correct when compared to those in the theory recalled in the paper.

**Reply by bcoconni** [reply]
> Going through the commit history I've figured out when this particular function was added and by whom.
> 
> "Complete rewrite of the FGInitialCondition class" - [da0b947](https://github.com/JSBSim-Team/jsbsim/commit/da0b947502b4a74788fed1cdeac3a60a95032c3e), in 2010, method originally called `SetAlphaRadIC()`.
> 
> The good news is the code author, @bcoconni is still around and very active in terms of JSBSim.

Oh man, that was many moons ago 🤔. I've found some notes I wrote at the time on this topic but there are not many details involved: I have most likely dumped the sheets on which I derived these equations. I'll need (quite) some time to wrap my head around these formulas and give a detailed explanation.

@emptyxiao, you might have found there are some errors in the results that these formulas give ? If that is so, I'd be interested in the cases you found that break the computations

**Reply by emptyxiao** [reply]
First of all, thank you for your reply！Recently, I was deriving the formulas in JSBSim. At present, this process stays at FGInitialCondition class .Learning FGTrim class is the next stage.
I have found a problem in function "UpdateCustomTransformMatrix()".I think it should use "mT.Transposed()"instead of "mT".I'm not sure I'm right.If you are also interesting in this problem.You may  check this function.
@bcoconni If you can give me a detailed explanation about  ‘’calcThetaBeta‘’,I would appreciate it very much.
